examples.md

```diff
@@ -657,3 +658,35 @@ steps:
       restore-keys: |
         ${{ runner.os }}-mint-
 ```
+
+## * - Bazel
+
+[`bazelisk`](https://github.com/bazelbuild/bazelisk) does not have be to separately downloaded and installed because it's already included in GitHub's `ubuntu-latest` and `macos-latest` base images.
+
+### Linux
+
+```yaml
+- name: Cache Bazel
+  uses: actions/cache@v3
+  with:
+    path: |
+      ~/.cache/bazel
+    key: ${{ runner.os }}-bazel-${{ hashFiles('.bazelversion', '.bazelrc', 'WORKSPACE', 'WORKSPACE.bazel', 'MODULE.bazel') }}
+    restore-keys: |
+      ${{ runner.os }}-bazel-
```

Comment:
@vsvipul I'm reluctant to make changes to a PR after a LGTM.. before I do, could you comment if you want this to be removed or not? Personally I'm not sure I see the harm in this being there, and in my project it does seem to help quite a bit to speed up things on small dependency changes.  It also does not seem consistent, to me, for Bazel to be the only example where there is not `restore-keys`? Isn't what @PiotrSikora alludes to something that could be said for all examples and that is not specific to Bazel per se?

Reply:
Skimming through examples, `deno` doesn't have `restore-keys` either.

But if you want to retain it here, I'd probably add more layers of `restore-keys`, e.g.:
```
restore-keys: |
    ${{ runner.os }}-bazel-${{ hashFiles('.bazelversion', '.bazelrc', 'WORKSPACE', 'WORKSPACE.bazel') }}
    ${{ runner.os }}-bazel-${{ hashFiles('.bazelversion', '.bazelrc') }}
    ${{ runner.os }}-bazel-${{ hashFiles('.bazelversion') }}
    ${{ runner.os }}-bazel-
```

Reply:
Its upto you @vorburger . But in most cases we don't include the hash in restore-keys the key with the hash is already accounted for in an exact match. Also, in some cases we don't keep restore-keys where the restored cache might not be very useful. I don't have much clue about bazel so can't comment on that. would leave the decision to you.

Reply:
Sorry, the `restore-keys` above is wrong, since those `hashFiles` would all evaluate to different hashes, even if the other files didin't change, and there would be no partial match. It should be this instead:
```
key: ${{ runner.os }}-bazel-${{ hashFiles('.bazelversion') }}-${{ hashFiles('.bazelrc') }}-${{ hashFiles('WORKSPACE', 'WORKSPACE.bazel') }}-${{ hashFiles('MODULE.bazel') }}
restore-keys: |
    ${{ runner.os }}-bazel-${{ hashFiles('.bazelversion') }}-${{ hashFiles('.bazelrc') }}-${{ hashFiles('WORKSPACE', 'WORKSPACE.bazel') }}
    ${{ runner.os }}-bazel-${{ hashFiles('.bazelversion') }}-${{ hashFiles('.bazelrc') }}
    ${{ runner.os }}-bazel-${{ hashFiles('.bazelversion') }}
    ${{ runner.os }}-bazel-
```

Reply:
why are different hashes required if all of them are expected to be present in a bazel project? wouldn't 1 suffice?

Reply:
Huh! I was sure that `hashFiles` hashes the _content_ of each matched file, so that the hash changes if the content changes (e.g. it would change if you added a new dependency to `MODULE.bazel`), but now that you've asked, I looked at the documentation again, and I'm not sure what it does anymore, since the documentation is a bit ambiguous:

> This function calculates an individual SHA-256 hash for each matched file, and then uses those hashes to calculate a final SHA-256 hash for the set of files.

Does it calculate SHA-256 of _the content_ or _the filename_ of the matched file?

Reply:
it is the content SHA i believe.

Reply:
Then I don't understand your previous question.

Using multiple `hashFiles` allows GHA to perform a partial match on the restore key, e.g. the first run had a key `linux-bazel-AAA-BBB-CCC-DDD-EEE`, but after modifying `WORKSPACE` file the new key is `linux-bazel-AAA-BBB-CCC-XXX-EEE`. With the `restore-keys` above, GHA would try to restore:
- `linux-bazel-AAA-BBB-CCC-XXX-EEE` - cache miss,
- `linux-bazel-AAA-BBB-CCC-XXX` - cache miss,
- `linux-bazel-AAA-BBB-CCC` - partial cache hit against `linux-bazel-AAA-BBB-CCC-DDD-EEE`.

If you were only using a single `hashFiles`, then a single change to any of those files would change the whole hash, ignoring existing caches (i.e. `linux-bazel-AAA` would change to `linux-bazel-XXX` in the example above).

Reply:
That is only useful if cache changes incrementally for all these different version files, and I believe that is not the case here. Separating the hashes will only increase complexity without any sufficient difference in performance i think. 

Also I would expect all of these to change together if some dependencies change `'.bazelversion', '.bazelrc', 'WORKSPACE', 'WORKSPACE.bazel', 'MODULE.bazel'` if that is not the case, why should bazelversion come first, then bazelrc and then other files, given that we are prefix matching.

Reply:
Piotr, I believe I now finally better understand what you are after. You mean that there is a sort of "hierarchy" involved here, yes? Like this, a .bazelversion change causes the entire cache to be thrown away (which would make sense to me as well, yes), but a .bazrlrc version causes some cache reuse, etc.? That's a nice idea - but I'm personally not familiar enough with the cache action to be sure that this will really work like this.  Vipul, thoughts?

Reply:
The thing is that i need to know when each of these files change `'.bazelversion', '.bazelrc', 'WORKSPACE', 'WORKSPACE.bazel', 'MODULE.bazel'` to allow some kind of reuse for cache and I have never used bazel myself so i'm hesitant to merge something which I'm not sure about. I feel the current state of the PR is good enough for normal users.

Reply:
@vsvipul can I do anything else to now see this get merged as-is?

Reply:
no. this lgtm.